src/panel_gwalker/_utils.py

```diff
@@ -66,12 +71,29 @@ def _infer_prop(s: pd.Series, i=None) -> dict:
     }
 
 
+SAMPLE_ROWS = 100
+
+
 @pn.cache(max_items=20, ttl=60 * 5, policy="LRU")
-def _raw_fields(data: pd.DataFrame | Dict[str, np.ndarray]) -> list[dict]:
-    if isinstance(data, dict):
-        return [_infer_prop(pd.Series(array, name=col)) for col, array in data.items()]
+def _raw_fields_core(data: pd.DataFrame) -> list[dict]:
+    return [_infer_prop(data[col], i) for i, col in enumerate(data.columns)]
+
+
+@nw.narwhalify
```

Comment:
Here I also used narwhals. Could be performance optimized. But I spent quite a lot of time just getting it working.